hikari/impl/bot.py

```diff
@@ -23,589 +23,481 @@
-        # <<inherited docstring from traits.EntityFactoryAware>>
+    def entity_factory(self) -> entity_factory_.EntityFactory:
         return self._entity_factory
 
     @property
-    def event_factory(self) -> event_factory_impl.EventFactoryImpl:
-        # <<inherited docstring from traits.EventFactoryAware>>
+    def event_factory(self) -> event_factory_.EventFactory:
         return self._event_factory
 
     @property
     def executor(self) -> typing.Optional[concurrent.futures.Executor]:
-        # <<inherited docstring from traits.ExecutorAware>>
         return self._executor
 
     @property
     def heartbeat_latencies(self) -> typing.Mapping[int, float]:
-        # <<inherited docstring from traits.ShardAware>>
-        return {shard_id: shard.heartbeat_latency for shard_id, shard in self._shards.items()}
+        return {s.id: s.heartbeat_latency for s in self._shards.values()}
 
     @property
     def heartbeat_latency(self) -> float:
-        # <<inherited docstring from traits.ShardAware>>
-        started_shards = [
-            shard.heartbeat_latency for shard in self._shards.values() if not math.isnan(shard.heartbeat_latency)
-        ]
-
-        if not started_shards:
-            return float("nan")
-
-        return sum(started_shards) / len(started_shards)
+        latencies = [s.heartbeat_latency for s in self._shards.values() if not math.isnan(s.heartbeat_latency)]
+        return sum(latencies) if latencies else float("nan")
 
     @property
     def http_settings(self) -> config.HTTPSettings:
-        # <<inherited docstring from traits.NetworkSettingsAware>>
         return self._http_settings
 
     @property
     def intents(self) -> typing.Optional[intents_.Intents]:
-        # <<inherited docstring from traits.ShardAware>>
         return self._intents
 
-    @property
-    def is_debug_enabled(self) -> bool:
-        """Return `builtins.True` if debugging is enabled.
-
-        Returns
-        -------
-        builtins.bool
-            `builtins.True` if debugging is enabled, `builtins.False` otherwise.
-
-        """
-        return self._debug
-
-    @property
-    def is_stateless(self) -> bool:
-        # <<inherited docstring from traits.CacheAware>>
-        return self._stateless
-
     @property
     def me(self) -> typing.Optional[users.OwnUser]:
-        # <<inherited docstring from traits.ShardAware>>
         return self._cache.get_me()
 
     @property
     def proxy_settings(self) -> config.ProxySettings:
-        # <<inherited docstring from traits.NetworkSettingsAware>>
         return self._proxy_settings
 
-    @property
-    def rest(self) -> rest_client_impl.RESTClientImpl:
-        # <<inherited docstring from traits.RESTAware>>
-        return self._rest
-
     @property
     def shards(self) -> typing.Mapping[int, gateway_shard.GatewayShard]:
-        # <<inherited docstring from traits.ShardAware>>
-        return self._shards
+        return types.MappingProxyType(self._shards)
```

Comment:
do we need to build a MappingProxyType every time we call this property? MappingProxyType should change as the mapping is changed upstream shouldn't it?

Reply:
how else would you suggest doing this? I'd rather not expose shards in a map directly as altering that can lead to other implications such as leaking of connections and resources, and preventing graceful shutdown.

Creating a mapping proxy type is trivial anyway, it exists on a C level as it is part of the conduit of how classes work, so I don't see an issue using it like this really.

It is on the level that if you were having performance issues, you shouldn't be repeatedly calling the property either.